pkg/certifier/clearlydefined/clearlydefined.go

```diff
@@ -88,139 +94,155 @@ func GetPkgDefinition(ctx context.Context, coordinate *coordinates.Coordinate) (
-	if err := json.Unmarshal(body, &definition); err != nil {
-		return nil, fmt.Errorf("error unmarshalling JSON: %v", err)
-	}
+		if genCDPkgDocs, err := generateDocument(definitionMap); err != nil {
+			return nil, fmt.Errorf("evaluateDefinitionForSource failed with error: %w", err)
+		} else {
+			generatedCDDocs = append(generatedCDDocs, genCDPkgDocs...)
+		}
 
-	if definition.Described.ReleaseDate == "" {
-		return nil, nil
+		if genCDSrcDocs, err := evaluateDefinitionForSource(ctx, client, definitionMap); err != nil {
+			return nil, fmt.Errorf("evaluateDefinitionForSource failed with error: %w", err)
+		} else {
+			generatedCDDocs = append(generatedCDDocs, genCDSrcDocs...)
+		}
 	}
-
-	return &definition, nil
+	return generatedCDDocs, nil
 }
 
 // CertifyComponent takes in the root component from the gauc database and does a recursive scan
 // to generate clearly defined attestations
 func (c *cdCertifier) CertifyComponent(ctx context.Context, rootComponent interface{}, docChannel chan<- *processor.Document) error {
-	logger := logging.FromContext(ctx)
-
 	packageNodes, ok := rootComponent.([]*root_package.PackageNode)
 	if !ok {
-		return ErrOSVComponentTypeMismatch
+		return ErrComponentTypeMismatch
 	}
 
-	packMap := map[string][]*root_package.PackageNode{}
+	var purls []string
 	for _, node := range packageNodes {
-		// skip any purls that are generated by GUAC as they will not be found in clearly defined
-		if strings.Contains(node.Purl, "pkg:guac") {
-			continue
-		}
-		if _, ok := packMap[node.Purl]; !ok {
-			coordinate, err := coordinates.ConvertPurlToCoordinate(node.Purl)
-			if err != nil {
-				logger.Errorf("failed to parse purl into coordinate with error: %v", err)
-				continue
-			}
-			definition, err := GetPkgDefinition(ctx, coordinate)
-			if err != nil {
-				return fmt.Errorf("failed get package definition from clearly defined with error: %w", err)
-			}
-			// if definition for the package is not found, continue to the next package
-			if definition == nil {
-				continue
-			}
-			if err := generateDocument(node.Purl, definition, docChannel); err != nil {
-				return fmt.Errorf("could not generate document from OSV results: %w", err)
-			}
-			packMap[node.Purl] = append(packMap[node.Purl], node)
-			if definition.Described.SourceLocation != nil {
-				srcDefinition, err := GetSrcDefinition(ctx, definition.Described.SourceLocation.Type, definition.Described.SourceLocation.Provider,
-					definition.Described.SourceLocation.Namespace, definition.Described.SourceLocation.Name, definition.Described.SourceLocation.Revision)
-				if err != nil {
-					return fmt.Errorf("failed get source definition from clearly defined with error: %w", err)
-				}
+		purls = append(purls, node.Purl)
+	}
 
-				// if definition for the source is not found, continue to the next package
-				if srcDefinition == nil {
-					continue
-				}
+	if genCDDocs, err := EvaluateClearlyDefinedDefinition(ctx, c.cdHTTPClient, purls); err != nil {
```

Comment:
code changes here cover batch querying and de- duplication of code from license scanner on ingestion.